tsl/src/nodes/skip_scan/exec.c

```diff
@@ -0,0 +1,316 @@
+
+static void
+skip_scan_begin(CustomScanState *node, EState *estate, int eflags)
+{
+	SkipScanState *state = (SkipScanState *) node;
+	state->ctx = AllocSetContextCreate(estate->es_query_cxt, "skipscan", ALLOCSET_DEFAULT_SIZES);
+
+	state->idx = (ScanState *) ExecInitNode(state->idx_scan, estate, eflags);
+	node->custom_ps = list_make1(state->idx);
+
+	if (IsA(state->idx_scan, IndexScan))
+	{
+		IndexScanState *idx = castNode(IndexScanState, state->idx);
+		state->scan_keys = &idx->iss_ScanKeys;
+		state->num_scan_keys = &idx->iss_NumScanKeys;
+		state->scan_desc = &idx->iss_ScanDesc;
+	}
+	else if (IsA(state->idx_scan, IndexOnlyScan))
+	{
+		IndexOnlyScanState *idx = castNode(IndexOnlyScanState, state->idx);
+		state->scan_keys = &idx->ioss_ScanKeys;
+		state->num_scan_keys = &idx->ioss_NumScanKeys;
+		state->scan_desc = &idx->ioss_ScanDesc;
+	}
+	else
+		elog(ERROR, "unknown subscan type in SkipScan");
+
+	/* scankeys are not setup for explain only */
+	if (eflags & EXEC_FLAG_EXPLAIN_ONLY)
+		return;
+
+	/* find position of our skip key
+	 * skip key is put as first key for the respective column in sort_indexquals
+	 */
+	ScanKey data = *state->scan_keys;
+	for (int i = 0; i < *state->num_scan_keys; i++)
+	{
+		if (data[i].sk_flags == SK_ISNULL && data[i].sk_attno == state->sk_attno)
+		{
+			state->skip_key = &data[i];
+			break;
+		}
+	}
+	if (!state->skip_key)
+		elog(ERROR, "ScanKey for skip qual not found");
+
+	if (has_nulls_first(state))
+		skip_scan_switch_stage(state, SS_NULLS_FIRST);
+	else
+		skip_scan_switch_stage(state, SS_NOT_NULL);
+}
+
+static bool
+has_nulls_first(SkipScanState *state)
+{
+	return state->nulls_first;
+}
+
+static bool
+has_nulls_last(SkipScanState *state)
+{
+	return !has_nulls_first(state);
+}
+
+static void
+skip_scan_rescan_index(SkipScanState *state)
+{
+	index_rescan(*state->scan_desc,
+				 *state->scan_keys,
+				 *state->num_scan_keys,
+				 NULL /*orderbys*/,
+				 0 /*norderbys*/);
+}
+
+/*
+ * Update skip scankey flags according to stage
+ */
+static void
+skip_scan_switch_stage(SkipScanState *state, SkipScanStage new_stage)
+{
+	Assert(new_stage > state->stage);
+
+	switch (new_stage)
+	{
+		case SS_NOT_NULL:
+			state->skip_key->sk_flags = SK_ISNULL | SK_SEARCHNOTNULL;
+			break;
+
+		case SS_VALUES:
+			state->skip_key->sk_flags = 0;
+			break;
+
+		case SS_NULLS_LAST:
+		case SS_NULLS_FIRST:
+			state->skip_key->sk_flags = SK_ISNULL | SK_SEARCHNULL;
+			state->skip_key->sk_argument = 0;
+			break;
+
+		default:
+			break;
+	}
+
+	state->stage = new_stage;
+}
+
+static void
+skip_scan_update_key(SkipScanState *state, TupleTableSlot *slot)
+{
+	if (!state->prev_is_null && !state->distinct_by_val)
+		pfree(DatumGetPointer(state->prev_datum));
+
+	MemoryContext old_ctx = MemoryContextSwitchTo(state->ctx);
+	state->prev_datum = slot_getattr(slot, state->distinct_col_attnum, &state->prev_is_null);
+	if (state->prev_is_null)
+	{
+		state->skip_key->sk_flags = SK_ISNULL;
```

Comment:
The skip_key->sk_flags should only be overwritten when we switch states (and already handled there) and this code sets the sk_argument.  I don't understand why skip_key->sk_flags is set to SK_ISNULL here. Why is this needed?

Reply:
SK_ISNULL is about the state of sk_argument, not setting SK_ISNULL here can lead to segfault for pointer datum types. but it's all theoretical since we should never arrive in that branch since the skip qual will prevent use from getting NULL values and we scan for NULL exlicitly